base/strings/types.jl

```diff
@@ -219,3 +219,44 @@ julia> "Test "^3
 
 pointer(x::SubString{String}) = pointer(x.string) + x.offset
 pointer(x::SubString{String}, i::Integer) = pointer(x.string) + x.offset + (i-1)
+
+
+"""
+    SecureString(string::AbstractString)
+
+A string where the contents will be securely wiped when the last reference to the secure
+string has been removed. If `string` is of type `String` then the memory of the original
+parameter will also be securely wiped.
+
+# Examples
+```jldoctest
+julia> str = "abc"
+"abc"
+
+julia> SecureString(str)
+"abc"
+
+julia> gc()  # Cause the SecureString above to be garbage collected
+
+julia> str
+"\0\0\0"
+```
+"""
+mutable struct SecureString <: AbstractString
+    string::String
```

Comment:
Vector{UInt8}?

Reply:
Yes, I need to switch this to use non-immutable memory.

Reply:
One nice thing I have noticed about using `String` is that when doing `SecureString("password")` the string literal memory will also be wiped out when running `securezero!(::SecureString)`.

Reply:
The above is also true when using `Vector{UInt8}` internally as vector can access the memory of the original `String`.

Reply:
> The above is also true when using Vector{UInt8} internally as vector can access the memory of the original String

Yes, but (1) you're not supposed to use such a vector to mutate a string, (2) we plan to fix this to least make it much harder to mutate a string via a vector.

Reply:
Wiping string literals is definitely not a good idea. It basically breaks your program. If secure data is put in a string literal there's nothing we can do about it.